Shopify/shopify-cli
# Support hot reload DX into theme app extensions

### WHY are these changes introduced?

This PR introduces the hot reload feature into the theme app extension (TAE) development experience.

Currently, theme app extensions developers need to perform repetitive operations in their development workflow, which is about:

1. Creating an app
2. Adding a theme app extension to a theme
3. Pushing the theme app extension
4. Refreshing the theme editor (repeat steps 3 and 4 during the development)

This effort aims to better support TAE developers, in steps 3 and 4, with a hot-reload mechanism that will prevent repetitive workflows.


### WHAT is this pull request doing?

This PR reuses the same components that support hot reload features in theme development (`shopify theme serve`), but in the context of theme app extension development (`shopify extension serve`):

- The **development server** - relies on the `Proxy`, the middleware stack, and the Shopify SFR to present theme app extensions in the host theme (and re-render parts of the template)

- The **syncer** - keeps the remote version of the theme app extension updated

- The **watcher** - triggers file system events when users change theme app extension files

### How to test your changes?

This PR introduces changes in the core components of `shopify theme serve` to make them flexible to support `shopify extension serve`. The, we might check both commands to avoid regressions and validate the new features.

#### shopify theme serve

- Create a theme and go to the theme directory
- Run `shopify-dev theme serve`
- Change a section file (e.g., `sections/announcement-bar.liquid`) and check the Google Chrome Console to validate the hot reload mechanism
- Change a layout file (e.g., `layout/theme.liquid`) and check the Google Chrome Console to validate the mechanism that reloads the entire page
- Validate the assets middleware by checking if the CLI is serving them locally
- Validate if the hot reload is working for CSS and JavaScript files
- Validate if the the CLI is pushing theme assets after the changes
- Delete the development theme and validate if the CLI creates a new theme after running `shopify-dev theme serve`
- Validate if the `shopify-dev theme serve` command is working with the  `-t, --theme` flag
- Validate if the `shopify-dev theme serve` command is working with the  `--port` flag

#### shopify extension serve

- Create an app, add an extension in the app, and go to the extension directory
- Run `shopify-dev extension serve`
- Change a block, a snippet, and an asset file and check the Google Chrome Console to validate the hot reload mechanism
- Validate the assets middleware by checking if the CLI is serving them locally
- Validate if the the CLI is pushing extensions assets after the changes (every 5s)
- Delete the host theme and validate if the CLI creates a new theme after running `shopify-dev theme serve`
- Validate if the `shopify-dev extension serve` command is working with the  `-t, --theme` flag
- Validate if the `shopify-dev extension serve` command is working with the  `--port` flag

![222](https://user-images.githubusercontent.com/1079279/186702636-775172b7-712b-4b07-97b6-6740ae427d99.gif)


### Post-release steps

This feature won't be offiacially supported on CLI 2.x, so no documentation is required.

However, after we merge this, we must open a PR on [shopify/cli](https://github.com/shopify/cli) (3.x) wrapping this command, and then update the 3.x documentation.

### Update checklist

- [x] (N/A) I've added a CHANGELOG entry for this PR (if the change is public-facing)
